### ISSUE TITLE
`ctypes` thread safety auditing (and fixing)

### ISSUE THREAD

**Issue by ZeroIntensity** [python/cpython#127945]
# Feature or enhancement

This is a tracking issue for all thread safety problems related to ctypes. I'll be working on this, but others can feel free to give me some help.

First of all, we need to find where the thread safety problems are.

```[tasklist]
### Auditing
- [x] Audit `_ctypes.c`
- [ ] Audit `_ctypes_test.c` (and probably the generated file too)
- [ ] Audit `callbacks.c`
- [ ] Audit `callproc.c`
- [ ] Audit `cfield.c`
- [x] Audit `malloc_closure.c`
- [ ] Audit `stgdict.c`
```

I'll be tracking the issues that get found when auditing here. The plan is to just create a new issue and link to it for each new problem instead of flooding this issue with PRs.

Generally, the workflow for fixes should follow most of the rules from #116738, but I suspect we'll need recursive mutexes for quite a few things related to callbacks, because it's difficult to tell what might be re-entrant, and we can't use critical sections for arbitrary function pointers.

```[tasklist]
### Known Issues
- [ ] https://github.com/python/cpython/issues/127946
- [ ] https://github.com/python/cpython/issues/128182
- [ ] https://github.com/python/cpython/issues/128485
- [ ] https://github.com/python/cpython/issues/128567
- [ ] https://github.com/python/cpython/issues/128570
- [ ] https://github.com/python/cpython/issues/131974
```

cc @encukou, as the ctypes genius, and @colesbury as the free-threading mastermind. 

### Linked PRs
* gh-131662
* gh-131710
* gh-131716
* gh-131896
* gh-131898
* gh-131899
* gh-132473
* gh-132552
* gh-132575
* gh-132646
* gh-132682
* gh-132720
* gh-132727
* gh-134332
* gh-134364

**Comment by encukou** [comment]
Thank you! I can review the ctypes parts :)

Heads-up: I'm currently [working in `cfield.c`/`PyCField*`](https://github.com/python/cpython/issues/128715); focus elsewhere to avoid conflicts.

**Comment by ZeroIntensity** [comment]
Got it, thanks! I'll probably just move to `callbacks.c` next.

**Comment by kumaraditya303** [comment]
I discussed this with Peter and moving forward I'll be taking over the thread safety work and continue it.

**Comment by kumaraditya303** [comment]
I investigated the thread safety of ctypes and here's the (incomplete) list of issues:
- #128485:  race in initialization of `dlmalloc` in `_ctypes_alloc_callback`. This is fixed by allocating a small memory block and freeing it in module initialization to force initialization  of the allocator and is now thread safe with https://github.com/python/cpython/pull/131633.
- the freelist in `malloc_closure.c` is not thread safe in free-threading. This can be fixed by adding lock around the alloc and free functions in free-threading in https://github.com/python/cpython/pull/131662.
- concurrent calls to `PyCStructUnionType_update_stginfo` needs to be made thread safe as it modifies the type's data. This can be fixed by adding critical section around reading and writing to stginfo data in free-threading as using just mutex can lead to re-entrancy deadlocks. 
- the initialization and concurrent modifications of stginfo is not thread safe
- the module level functions such as `resize` are not thread safe and need to acquire critical section of the object being resized.

**Comment by colesbury** [comment]
I think there are also races on `StgInfo.flags`, in particular with the `DICTFLAG_FINAL` bit.

https://github.com/python/cpython/blob/4596666a9f9f6b97085619553a5cecb84ff6486d/Modules/_ctypes/_ctypes.c#L718-L719

**Comment by kumaraditya303** [comment]
> I think there are also races on StgInfo.flags, in particular with the DICTFLAG_FINAL bit.

Yes, I am covering that under "the initialization and concurrent modifications of stginfo is not thread safe".

**Comment by colesbury** [comment]
This is different and more important than concurrent modification or initialization of stginfo, and I think it should be one of the early things we should focus on. Concurrent modifications of ctypes Structures is not an important use case.

However, creating separate instances of the same structure is an important use case:

```python
from ctypes import Structure, c_int
from concurrent.futures import ThreadPoolExecutor

class POINT(Structure):
    _fields_ = [("x", c_int), ("y", c_int)]

with ThreadPoolExecutor() as executor:
    futures = [executor.submit(POINT, i, i) for i in range(1000)]
    results = [future.result() for future in futures]
```

This needs to be both efficient and thread-safe.

**Comment by sergey-miryanov** [comment]
Can this issue be closed now as all linked PRs have been merged? Is there anything else that needs to be done?
